overlord/snapstate/backend/snapdata.go

```diff
@@ -68,19 +68,27 @@ func (b Backend) RemoveSnapSaveData(snapInfo *snap.Info, dev snap.Device) error
 	return os.RemoveAll(saveDir)
 }
 
-// RemoveSnapDataDir removes base snap data directory
-func (b Backend) RemoveSnapDataDir(info *snap.Info, hasOtherInstances bool) error {
+// RemoveSnapDataDir removes base snap data directories.
+func (b Backend) RemoveSnapDataDir(info *snap.Info, hasOtherInstances bool, opts *dirs.SnapDirOptions) error {
 	if info.InstanceKey != "" {
 		// data directories of snaps with instance key are never used by
 		// other instances
-		if err := os.Remove(snap.BaseDataDir(info.InstanceName())); err != nil && !os.IsNotExist(err) {
+		dirs, err := snapBaseDataDirs(info.InstanceName(), opts)
+		if err != nil {
+			return err
+		}
+		if err = removeDirs(dirs); err != nil {
 			return fmt.Errorf("failed to remove snap %q base directory: %v", info.InstanceName(), err)
 		}
 	}
 	if !hasOtherInstances {
 		// remove the snap base directory only if there are no other
 		// snap instances using it
-		if err := os.Remove(snap.BaseDataDir(info.SnapName())); err != nil && !os.IsNotExist(err) {
```

Comment:
Maybe I'm overly careful here but I notice the semantic of the functions here is changing, the old code (that is removed here) is doing a `os.Remove()` which means that non-empty dirs will fail. The new code is doing a `os.RemoveAll()` which is different. Would things still work if we keep the "normal" os.Remove() here?

Reply:
Great point! Actually no, It wouldn't work without `RemoveAll` because the current symlink is left behind for snap user data [#2009617](https://bugs.launchpad.net/snapd/+bug/2009617).
I tried handling the deletion of current symlink separately, but I found that the per-user current symlinks (and data btw) are populated lazily in [snap run](https://github.com/snapcore/snapd/blob/00643e32a73521b526de0cd8430b80511a647b19/cmd/snap/cmd_run.go#L1045)'s first run. unlike `/var/snap/$SNAP` current symlink is created/updated/removed in `backend.LinkSnap` and `backend.UnlinkSnap`.